tool.c

```diff
@@ -171,6 +171,72 @@ commands(int argc, char *argv[])
 	return ret;
 }
 
+static void
+graphviznode(struct node *n)
+{
+	size_t i;
+	char *order_only;
+	struct edge *e;
+
+	printf("\"%p\" [label=\"%s\"]\n", (void *)n, n->path->s);
+
+	e = n->gen;
+	if (!e || (e->flags & FLAG_WORK))
+		return;
+	e->flags |= FLAG_WORK;
+
+	for (i = 0; i < e->nin; i++)
+		graphviznode(e->in[i]);
+
+	if (e->nin == 1 && e->nout == 1) {
+		printf("\"%p\" -> \"%p\" [label=\" %s\"]\n", (void *)e->in[0], (void *)e->out[0], e->rule->name);
+	} else {
+		printf("\"%p\" [label=\"%s\", shape=ellipse]\n", (void *)e, e->rule->name);
+		for (i = 0; i < e->nout; i++)
+			printf("\"%p\" -> \"%p\"\n", (void *)e, (void *)e->out[i]);
+		for (i = 0; i < e->nin; i++) {
+			order_only = "";
+			if (i >= e->inorderidx)
+				order_only = " style=dotted";
+			printf("\"%p\" -> \"%p\" [arrowhead=none%s]\n", (void *)e->in[i], (void *)e, order_only);
+		}
+	}
+}
+
+static int
+graphviz(int argc, char *argv[])
```

Comment:
I'd call this `graph` (even though it is a bit generic), so that the function name always matches the tool name.

Reply:
Also renamed the recursive node function